### PR TITLE
Bump support annotations to 27.1.0.

### PR DESCRIPTION
**Pull request by naturalwarren**

### CODE REVIEW

**Comment by naturalwarren** [comment]
I've updated the tests.

I'll open a separate PR for `LongDef` support since that's a new piece of functionality we can build on-top of this bump. Both this change a `LongDef` support will go out in the same release.

**Comment by hzsweers** [comment]
I don't think you've covered all of them. When I started on this in #53, I was at this point before taking pause. I don't see how you can get by without updating the many samples and test models.

screen shot 2018-03-27 at 6 39 06 pm

File: rave-compiler/src/test/java/com/uber/rave/model/IntRangeTestModel.java [hunk 1]
```diff
@@ -14,31 +14,31 @@
  */
 @Validated(factory = MyFactory.class)
 public class IntRangeTestModel {
-    private static final long FROM = -15L;
-    private static final long TO = 1000L;
+    private static final int FROM = -15;
```
Comment (naturalwarren) [inline rave-compiler/src/test/java/com/uber/rave/model/IntRangeTestModel.java]: I think we can actually leave these since `IntRange` defaults to longs. Thoughts?

File: rave-test/src/main/java/com/uber/rave/AnnotationSpecs.java [hunk 2]
```diff
@@ -199,7 +199,7 @@ public Builder setIntDef(long... values) {
          * @param to the upper bound of the rnage inclusive.
          * @return this {@link Builder}
          */
-        public Builder setIntRange(long from, long to) {
+        public Builder setIntRange(int from, int to) {
```
Comment (naturalwarren) [inline rave-test/src/main/java/com/uber/rave/AnnotationSpecs.java]: Ditto here...

**Comment by naturalwarren** [comment]
Strange that this test is failing, it's not on my branch. `testFieldMethodValidationGeneration_whenFieldsArePrivate_shouldGenerateAppropriateMethods`

File: rave-compiler/src/test/java/com/uber/rave/model/IntRangeTestModel.java (hunk 1 above)
Comment (naturalwarren) [reply rave-compiler/src/test/java/com/uber/rave/model/IntRangeTestModel.java]: I've posted in ASG if a similar change is going to be made to IntRange lets see what Ian says.

Comment (naturalwarren) [reply rave-compiler/src/test/java/com/uber/rave/model/IntRangeTestModel.java]: I actually like this better that supporting longs and I suspect IntRange will eventually move to the same thing per the problems described with IntDef here:

https://android.googlesource.com/platform/frameworks/support/+/df64f2ab05e32519d95adc1625cdd840f650956c

**Comment by hzsweers** [comment]
Ping when tests are ready

**Comment by naturalwarren** [comment]
Yeah I'm not sure why travis is failing but the code should be complete here. I think its rebasing my change onto master...

**Comment by naturalwarren** [comment]
@hzsweers Tests are ready.

File: build.gradle [hunk 3]
```diff
@@ -16,7 +16,7 @@ subprojects {
 }
 
 ext.deps = [
-        supportAnnotations: 'com.android.support:support-annotations:23.0.0',
+        supportAnnotations: 'com.android.support:support-annotations:27.1.0',
```
Comment (hzsweers) [inline build.gradle]: 27.1.1 while we're at it

File: rave-test/src/main/java/com/uber/rave/DoubleCreator.java [hunk 4]
```diff
@@ -26,11 +26,6 @@
  */
 public class DoubleCreator extends ObjectCreator<Double> {
 
-
-    public DoubleCreator(long... validValues) {
```
Comment (hzsweers) [inline rave-test/src/main/java/com/uber/rave/DoubleCreator.java]: This seems unrelated

File: rave-test/src/main/java/com/uber/rave/IntCreator.java [hunk 5]
```diff
@@ -27,16 +27,16 @@
  * This class is creates valid and invalid permutations of a long argument based on the annotation
  * restrictions.
  */
-public final class LongCreator extends ObjectCreator<Long> {
```
Comment (hzsweers) [inline rave-test/src/main/java/com/uber/rave/IntCreator.java]: Shouldn't you keep this for LongDef support in the future? Let's create a new `IntCreator` rather than introduce churn here. Also the docs on this were not updated anyway

File: sample/build.gradle [hunk 6]
```diff
@@ -11,13 +11,13 @@ buildscript {
 apply plugin: 'com.android.application'
 
 android {
-    compileSdkVersion 25
+    compileSdkVersion 27
     buildToolsVersion "25.0.2"
```
Comment (hzsweers) [inline sample/build.gradle]: This should match your compile sdk

File: build.gradle (hunk 3 above)
Comment (naturalwarren) [reply build.gradle]: `41c9d20`

File: rave-test/src/main/java/com/uber/rave/DoubleCreator.java (hunk 4 above)
Comment (naturalwarren) [reply rave-test/src/main/java/com/uber/rave/DoubleCreator.java]: This wasn't used and stopped working since intDef can no longer take `long` values so I deleted it.

File: sample/build.gradle (hunk 6 above)
Comment (naturalwarren) [reply sample/build.gradle]: Ah good catch. `136a638`

File: rave-test/src/main/java/com/uber/rave/IntCreator.java (hunk 5 above)
Comment (naturalwarren) [reply rave-test/src/main/java/com/uber/rave/IntCreator.java]: I was thinking we'd introduce LongCreator as part of LongDef support. Introducing a new `IntCreator` requires migrating old call sites since we've made a breaking change in the constructor. 

Updated the doc at the top of this class in `28172a3`. Let me know if that's cool :)

File: build.gradle (hunk 3 above)
Comment (naturalwarren) [reply build.gradle]: 41c9d20

File: sample/build.gradle (hunk 6 above)
Comment (naturalwarren) [reply sample/build.gradle]: Ah good catch. 1136a638

File: rave-test/src/main/java/com/uber/rave/DoubleCreator.java (hunk 4 above)
Comment (hzsweers) [reply rave-test/src/main/java/com/uber/rave/DoubleCreator.java]: How did it stop working if it wasn't used?

Comment (naturalwarren) [reply rave-test/src/main/java/com/uber/rave/DoubleCreator.java]: Compile time breaking change. `AnnotationSpecs.setIntDef` no longer takes long values.

Comment (hzsweers) [reply rave-test/src/main/java/com/uber/rave/DoubleCreator.java]: So was is `This wasn't used` or `compile time breaking change`? I'm a little confused at this point, especially since you added `LongDef` back now and presumably would use this?

Comment (naturalwarren) [reply rave-test/src/main/java/com/uber/rave/DoubleCreator.java]: > So was is This wasn't used or compile time breaking change?

Both. We had to change AnnotationSpecs.setIntDef to only accept ints which broke this constructor at compile time. Since the constructor was unused but causing compilation to fail I opted just to remove it.

File: rave-compiler/src/test/resources/fixtures/test2/MyFactory_Generated_Validator.java [hunk 7]
```diff
@@ -35,6 +36,7 @@
     addSupportedClass(ArrayNotNull.class);
     addSupportedClass(AbstractAnnotated.class);
     addSupportedClass(IntDefModel.class);
+    addSupportedClass(LongDefModel.class);
```
Comment (hzsweers) [inline rave-compiler/src/test/resources/fixtures/test2/MyFactory_Generated_Validator.java]: Alpha this like the rest. I know it's generated code, so you'll want to fix in the generation

File: rave-test/src/main/java/com/uber/rave/AnnotationSpecs.java [hunk 8]
```diff
@@ -192,6 +203,18 @@ public Builder setIntDef(int... values) {
...
         }
 
+        /**
+         * Set the valid values for the long in a {@link LongDef}.
+         *
+         * @param values the valid values.
+         * @return this {@link Builder}
+         */
+        public Builder setLongDef(long ... values) {
+            annotationSpecs.hasLongDef= true;
```
Comment (hzsweers) [inline rave-test/src/main/java/com/uber/rave/AnnotationSpecs.java]: Please fix the code style in this method

File: rave/src/main/java/com/uber/rave/BaseValidator.java [hunk 9]
```diff
@@ -523,14 +523,37 @@ protected static ValidationContext getValidationContext(Class<?> clazz) {
...
+     * we may add support later.
+     *
+     * @param validationContext The context of the item in the class being validated. This is used in case of an error.
+     * @param value The long value to check.
+     * @param flag the is the flag value from IntDef it is currently unsupported.
+     * @param acceptableValues the values that the value parameter can take on.
+     * @return an error if value is not present in list of acceptable values. Otherwise returns an empty list.
+     */
+    protected static List<RaveError> checkLongDef(
+            ValidationContext validationContext, long value, boolean flag,
```
Comment (hzsweers) [inline rave/src/main/java/com/uber/rave/BaseValidator.java]: If at least two params are on different lines, put all on separate lines

File: rave/src/main/java/com/uber/rave/BaseValidator.java [hunk 10]
```diff
@@ -689,6 +712,28 @@ protected static ValidationContext getValidationContext(Class<?> clazz) {
...
+            } else {
+                stringBuilder.append("{");
+            }
+            stringBuilder.append(string);
+            first = false;
+        }
+        stringBuilder.append("}");
+
+        return createNewList(
+                new RaveError(validationContext, value + " " + RaveErrorStrings.LONG_DEF_ERROR
```
Comment (hzsweers) [inline rave/src/main/java/com/uber/rave/BaseValidator.java]: Use `String.format` rather than hard-to-read concatenation

File: rave-compiler/src/main/java/com/uber/rave/compiler/RaveProcessor.java [hunk 11]
```diff
@@ -247,8 +248,8 @@ private void addImplicitNullnessAnnotations(
     }
 
     /**
-     * This method extracts Def type annotations. These include {@link StringDef} {@link IntDef}. These annotations
-     * are extended so there is a extra level of annotation hierarchy traversal that is required.
+     * This method extracts Def type annotations. These include {@link StringDef}, {@link IntDef} and {@link LongDef}.
```
[the remark this answers is not in the document]
Comment (behroozkhorashadi) [reply rave-compiler/src/main/java/com/uber/rave/compiler/RaveProcessor.java]: This comment isn't useful. Can we just just rephrase the portion of the sentence that we need to change.

File: rave-compiler/src/test/java/com/uber/rave/compiler/CompilerIntegrationTest.java [hunk 12]
```diff
@@ -176,6 +177,33 @@ public void allInvalidIntDefModel_shouldFail() throws RaveException {
...
+        Collection<LongDefModel> validCases = builder.getInvalidCases();
+        assertFalse(validCases.isEmpty());
+        Rave rave = Rave.getInstance();
+        for (LongDefModel model : validCases) {
+            try {
+                rave.validate(model);
+            } catch (InvalidModelException e) {
+                continue;
+            }
+            assertTrue("Model should have failed with int: " + model.getStandard() + "\n", false);
```
[the remark this answers is not in the document]
Comment (behroozkhorashadi) [reply rave-compiler/src/test/java/com/uber/rave/compiler/CompilerIntegrationTest.java]: Every one of these models should throw a InvalidModelException so we should never reach this line.

File: rave-compiler/src/test/resources/fixtures/longdef/BadLongDefUse.java [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+
+import com.uber.rave.annotation.Validated;
+
+import fixtures.SampleFactory;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import android.support.annotation.LongDef;
+
+@Validated(factory = SampleFactory.class)
+public class BadLongDefUse {
```
[the remark this answers is not in the document]
Comment (behroozkhorashadi) [reply rave-compiler/src/test/resources/fixtures/longdef/BadLongDefUse.java]: I think there is a test that already does that. @naturalwarren can you double check

File: rave/src/main/java/com/uber/rave/BaseValidator.java (hunk 9 above)
Comment (behroozkhorashadi) [reply rave/src/main/java/com/uber/rave/BaseValidator.java]: +1

File: rave/src/main/java/com/uber/rave/BaseValidator.java (hunk 10 above)
Comment (behroozkhorashadi) [reply rave/src/main/java/com/uber/rave/BaseValidator.java]: +1

**Comment by behroozkhorashadi** [comment]
I think we should try an make some progress on this. I know that we want to maintain the quality of this open source repo but perfect is the enemy of good.

File: rave-compiler/src/test/resources/fixtures/longdef/BadLongDefUse.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (naturalwarren) [reply rave-compiler/src/test/resources/fixtures/longdef/BadLongDefUse.java]: Test that covers that case is `RaveUnitTest.checkLongDef_whenValueDoesNotMatch_shouldProduceError`

File: rave-test/src/main/java/com/uber/rave/AnnotationSpecs.java (hunk 8 above)
Comment (naturalwarren) [reply rave-test/src/main/java/com/uber/rave/AnnotationSpecs.java]: Good catch. dfa33e7

File: rave-compiler/src/test/resources/fixtures/test2/MyFactory_Generated_Validator.java (hunk 7 above)
Comment (naturalwarren) [reply rave-compiler/src/test/resources/fixtures/test2/MyFactory_Generated_Validator.java]: We're actually not alpha here, peak a little bit above :) Would have to add a sort by class name in the processor and switch all the tests.

File: rave-compiler/src/main/java/com/uber/rave/compiler/AnnotationWriter.java [hunk 14]
```diff
@@ -157,6 +159,19 @@ void write(@Nullable IntDef intDef) {
         buildStatements(baseWriter.getFormattedString(), baseWriter.getArgs());
     }
 
+    void write(@Nullable LongDef longDef) {
+        checkAnnotationNotNull(longDef);
```
[the remark this answers is not in the document]
Comment (naturalwarren) [reply rave-compiler/src/main/java/com/uber/rave/compiler/AnnotationWriter.java]: Yeah, not sure. Just modeled what writing an IntDef was doing. Perhaps @behroozkhorashadi can comment here.

File: rave/src/main/java/com/uber/rave/BaseValidator.java (hunk 9 above)
Comment (naturalwarren) [reply rave/src/main/java/com/uber/rave/BaseValidator.java]: Good call. 8f4fd10

File: rave-compiler/src/test/resources/fixtures/longdef/BadLongDefUse.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (hzsweers) [reply rave-compiler/src/test/resources/fixtures/longdef/BadLongDefUse.java]: I meant in this model fixture

File: rave-compiler/src/main/java/com/uber/rave/compiler/RaveProcessor.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (hzsweers) [reply rave-compiler/src/main/java/com/uber/rave/compiler/RaveProcessor.java]: Can you address Behrooz's comment?

File: rave-compiler/src/test/java/com/uber/rave/compiler/CompilerIntegrationTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (hzsweers) [reply rave-compiler/src/test/java/com/uber/rave/compiler/CompilerIntegrationTest.java]: Second part of my comment still stands

> Seems like what you want to do is throw an AssertionError here with that message rather than a weird assertion of a false boolean

File: rave-compiler/src/test/resources/fixtures/longdef/BadLongDefUse.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (naturalwarren) [reply rave-compiler/src/test/resources/fixtures/longdef/BadLongDefUse.java]: Ah gotcha. Similar fixtures like `BadIntDefUse` haven't included that case in the past but it seems worthwhile since it would server as an end-to-end test that ensure the internals of RAVE are properly hooked up which I don't think we've covered anywhere else.

File: rave-compiler/src/main/java/com/uber/rave/compiler/RaveProcessor.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (naturalwarren) [reply rave-compiler/src/main/java/com/uber/rave/compiler/RaveProcessor.java]: I think this is good, I added an Oxford comma. Is there something else reviewers are looking for here? @behroozkhorashadi

File: rave-compiler/src/test/java/com/uber/rave/compiler/CompilerIntegrationTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (naturalwarren) [reply rave-compiler/src/test/java/com/uber/rave/compiler/CompilerIntegrationTest.java]: Yeah, I think we can improve this. I'll make the change.

File: rave-compiler/src/test/resources/fixtures/longdef/BadLongDefUse.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (naturalwarren) [reply rave-compiler/src/test/resources/fixtures/longdef/BadLongDefUse.java]: Looking a little bit more closely we've covered this in `CompilerIntegrationTest.allInvalidLongDefModel_shouldFail()`. In this test construct a `LongDefModel` which returns a value that isn't part of the `LongDef`.

File: rave-compiler/src/test/java/com/uber/rave/compiler/CompilerIntegrationTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (naturalwarren) [reply rave-compiler/src/test/java/com/uber/rave/compiler/CompilerIntegrationTest.java]: 9b68ad3
